include/Gaffer/ShufflePlug.inl

```diff
@@ -102,26 +103,28 @@ T ShufflesPlug::shuffle( const T &sourceContainer ) const
 				const std::string destination = scope.context()->substitute( plug->destinationPlug()->getValue() );
 				if( !destination.empty() && IECore::StringAlgo::match( sourceData.first, shuffle.sourcePattern ) )
 				{
-					destinationContainer[destination] = sourceData.second;
-					if( !newDestinations.insert( destination ).second )
+					if( shuffle.overwrite || sourceContainer.find( destination ) == sourceContainer.end() )
 					{
-						throw IECore::Exception(
-							boost::str(
-								boost::format(
-									"ShufflesPlug::shuffle : Destination plug \"%s\" is shuffling \"%s\" to \"%s\", " \
-									"but this destination was already written by a previous Shuffle."
-								) %
-								plug->destinationPlug()->relativeName( plug->node() ? plug->node()->parent() : nullptr ) %
-								sourceData.first %
-								destination
-							)
-						);
+						destinationContainer[destination] = sourceData.second;
+						if( !newDestinations.insert( destination ).second )
+						{
+							throw IECore::Exception(
+								boost::str(
+									boost::format(
+										"ShufflesPlug::shuffle : Destination plug \"%s\" is shuffling \"%s\" to \"%s\", " \
+										"but this destination was already written by a previous Shuffle."
+									) %
+									plug->destinationPlug()->relativeName( plug->node() ? plug->node()->parent() : nullptr ) %
+									sourceData.first %
+									destination
+								)
+							);
```

Comment:
I can't comment on the history of why it was implemented as O(n^2) so I have just left it that way. This commit adds a map lookup as well on line 106 but that should be O(log(n)).

I left the exception as it seems appropriate. Its checking for the case where the user has specified multiple shuffles writing to the same destination name, which is not what the overwrite option is adding. It is adding the capability to overwrite (or not) existing source data with the same name as the current shuffle's destination, not overwrite already written destination data with the same name as the current shuffle's destination.

Reply:
I think the subtlety here is that data from the source container is written to the destination container in two different ways.
1. A shuffle uses the name of the data as its source. In this case the data is copied to the destination container and has the shuffle's destination name.
2. No shuffle uses the name of the data as its source. In this case the data is copied to the destination container and has the same name as it did in the source container.

The current behaviour is that shuffles take precedence over straight copies. So if we have source data, `{ "P", "Pref", "N" }` and a shuffle `"P" -> "Pref"`, the copy of "Pref" to "Pref" does not happen, because there is a shuffle writing to "Pref". The original "Pref" data in the source is being overwritten/replaced by the result of the shuffle. This PR adds the option to not overwrite/replace existing source data with the same name as the destination of a shuffle, instead the copy of the original source data will happen.

The other case is that there is more than one shuffle with the same destination name. So if we have source data, `{ "P", "Pref", "N" }` and shuffles, `"P" -> "Pref"` and `"N" -> "Pref"` then the copy of "Pref" to "Pref" does not currently happen. But the code detects that there are two shuffles with the same destination name and throws an exception. The alternative to throwing an exception would be to combine the source data from the two shuffles in some way (which as you mention seems out-of-scope for this PR).

I suppose with this PR if there are multiple shuffles that write to the same destination name, but their overwrite/replace flags are set to false, what should happen? I think the code in this PR as it is means that there will be no exception and the copy will happen. E.g. source data `{ "P", "Pref", "N" }` and shuffles `"P" -> "Pref" (replace = false)` and `"N" -> "Pref" (replace = false)`, neither shuffle will happen and the copy "Pref" -> "Pref" will happen with no exception. This seems wrong, I can add a test case for this and ensure that an exception occurs whenever there are multiple shuffles that specify the same destination name. However this PR as it is matches the current behaviour with regard to the enabled flag. So if there are multiple shuffles with the same destination name but they are disabled the shuffles do not happen, the copy happens and there is no exception.

Reply:
I have pushed an additional commit that refactors the shuffle function so that shuffles are performed in order of definition and separate shuffles may write to the same destination name. If a shuffle's source contains no wildcards then the shuffle amounts to one move, however when the source contains wildcards the shufffle may amount to a set of moves. Each shuffle's set of moves is unordered so moves in the same set cannot write to the same destination.